From 5cd00b069b1074d24d09d91b3e10c1c8e1ca8cc0 Mon Sep 17 00:00:00 2001
From: Hai Zhang <zhanghai@google.com>
Date: Tue, 12 Oct 2021 19:54:17 -0700
Subject: [PATCH] Default overrideUserWhenGranting to false.

The previous default value of true made sense when most of the roles
were default apps, and the user action of setting a default app
implies that the permissions should always be given to make the
default app work. However since nowadays most of the roles are for
pregrants, these pregrant roles should never override user's explicit
denial, and the previous default apps are marked as so explicitly
instead.

In other words, this CL changes all the non-defaults app roles to never
override user set permissions, but keeps allowing default app roles to
do so.

Fixes: 202312327
Test: presubmit
Change-Id: I8b862eebf61acf4cbb1399c9869054a1414a2394
Merged-In: I8b862eebf61acf4cbb1399c9869054a1414a2394
(cherry picked from commit ce756728482bae444b71d4b3a016218c65f8cdb4)
(cherry picked from commit a360d5f391d02be49a2dbf46e912fe0b2155f9fd)
Merged-In:I8b862eebf61acf4cbb1399c9869054a1414a2394
---
 PermissionController/res/xml/roles.xml                   | 9 ++++++++-
 .../permissioncontroller/role/model/RoleParser.java      | 2 +-
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/PermissionController/res/xml/roles.xml b/PermissionController/res/xml/roles.xml
index 583cdcbfc..30052ad49 100644
--- a/PermissionController/res/xml/roles.xml
+++ b/PermissionController/res/xml/roles.xml
@@ -91,6 +91,7 @@
         fallBackToDefaultHolder="true"
         showNone="true"
         label="@string/role_assistant_label"
+        overrideUserWhenGranting="true"
         requestable="false"
         shortLabel="@string/role_assistant_short_label">
         <required-components>
@@ -143,6 +144,7 @@
         description="@string/role_browser_description"
         exclusive="true"
         label="@string/role_browser_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_browser_request_description"
         requestTitle="@string/role_browser_request_title"
         shortLabel="@string/role_browser_short_label">
@@ -183,6 +185,7 @@
         exclusive="true"
         fallBackToDefaultHolder="true"
         label="@string/role_dialer_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_dialer_request_description"
         requestTitle="@string/role_dialer_request_title"
         searchKeywords="@string/role_dialer_search_keywords"
@@ -257,6 +260,7 @@
         description="@string/role_sms_description"
         exclusive="true"
         label="@string/role_sms_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_sms_request_description"
         requestTitle="@string/role_sms_request_title"
         searchKeywords="@string/role_sms_search_keywords"
@@ -344,6 +348,7 @@
         description="@string/role_emergency_description"
         exclusive="true"
         label="@string/role_emergency_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_emergency_request_description"
         requestTitle="@string/role_emergency_request_title"
         searchKeywords="@string/role_emergency_search_keywords"
@@ -373,6 +378,7 @@
         description="@string/role_home_description"
         exclusive="true"
         label="@string/role_home_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_home_request_description"
         requestTitle="@string/role_home_request_title"
         searchKeywords="@string/role_home_search_keywords"
@@ -409,6 +415,7 @@
         description="@string/role_call_redirection_description"
         exclusive="true"
         label="@string/role_call_redirection_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_call_redirection_request_description"
         requestTitle="@string/role_call_redirection_request_title"
         shortLabel="@string/role_call_redirection_short_label"
@@ -429,6 +436,7 @@
         description="@string/role_call_screening_description"
         exclusive="true"
         label="@string/role_call_screening_label"
+        overrideUserWhenGranting="true"
         requestDescription="@string/role_call_screening_request_description"
         requestTitle="@string/role_call_screening_request_title"
         shortLabel="@string/role_call_screening_short_label"
@@ -486,7 +494,6 @@
         description="@string/role_watch_description"
         exclusive="false"
         minSdkVersion="31"
-        overrideUserWhenGranting="false"
         systemOnly="false"
         visible="false">
         <permissions>
diff --git a/PermissionController/src/com/android/permissioncontroller/role/model/RoleParser.java b/PermissionController/src/com/android/permissioncontroller/role/model/RoleParser.java
index 173a28f92..b3a02528e 100644
--- a/PermissionController/src/com/android/permissioncontroller/role/model/RoleParser.java
+++ b/PermissionController/src/com/android/permissioncontroller/role/model/RoleParser.java
@@ -349,7 +349,7 @@ public class RoleParser {
                 Build.VERSION_CODES.BASE);
 
         boolean overrideUserWhenGranting = getAttributeBooleanValue(parser,
-                ATTRIBUTE_OVERRIDE_USER_WHEN_GRANTING, true);
+                ATTRIBUTE_OVERRIDE_USER_WHEN_GRANTING, false);
 
         boolean requestable = getAttributeBooleanValue(parser, ATTRIBUTE_REQUESTABLE, visible);
         Integer requestDescriptionResource;
-- 
2.33.1

